homeassistant/components/bryant_evolution/coordinator.py

```diff
@@ -0,0 +1,190 @@
+"""DataUpdateCoordinator for the bryant_evolution module."""
+
+from dataclasses import dataclass
+from datetime import timedelta
+import logging
+
+from evolutionhttp import BryantEvolutionLocalClient
+
+from homeassistant.components.climate import HVACAction, HVACMode
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class _ZoneState:
+    """Cached state of an Evolution zone."""
+
+    curr_temp: int
+    htsp: int
+    clsp: int
+
+
+@dataclass(frozen=True)
+class _SystemState:
+    """Cached state of a single Evolution system."""
+
+    fan_mode: str
+    hvac_mode: str
+    is_hvac_active: bool
+    zones: dict[int, _ZoneState]
+
+
+@dataclass(frozen=True)
+class EvolutionState:
+    """Class to hold the cached state of all Evolution systems accessible through one SAM device."""
+
+    systems: dict[int, _SystemState]
+
+
+async def _read_evolution_state(
+    system_zones: list[tuple[int, int]], tty: str
+) -> EvolutionState:
+    """Read the state of the given systems and zones from the device at tty."""
+
+    # Convert the input list of (system_id, zone_id) to a dict[system_id, list[zone_id]]
+    system_to_zones: dict[int, list[int]] = {}
+    for system_id, zone_id in system_zones:
+        system_to_zones.setdefault(system_id, []).append(zone_id)
+
+    # Read each system's state.
+    system_states: dict[int, _SystemState] = {}
+    for sys_id, zone_ids in system_to_zones.items():
+        system_states[sys_id] = await _read_system_state(sys_id, zone_ids, tty)
+    return EvolutionState(systems=system_states)
+
```

Comment:
I'd prefer to keep a domain-typed return value; see https://testing.googleblog.com/2017/11/obsessed-with-primitives.html for motivation.

That said, per one of your other comments, I think the accessors for the state don't make sense to keep on the coordinator. But, rather than putting them or helpers/utils, we should keep them closely coupled with the data on which they act. Moved them here.